Pear/Controller/ChatViewController.swift

```diff
@@ -240,16 +244,57 @@ class ChatViewController: UIViewController {
     }
     
     @objc func keyboardWillShow(notification: NSNotification) {
-        if let keyboardSize = (notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue {
-            if view.frame.origin.y == 0 {
-                view.frame.origin.y -= keyboardSize.height
+        if let keyboardSize = (notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue, let keyboardDuration = notification.userInfo?[UIResponder.keyboardAnimationDurationUserInfoKey] as? Double {
+            if keyboardOffSet == 0 {
+                keyboardOffSet = chatInputContainerView.frame.maxY - keyboardSize.minY
+                chatInputContainerView.snp.updateConstraints { make in
+                    make.bottom.equalToSuperview().inset(40+keyboardOffSet)
```

Comment:
since all of these keyboard functions are repeating the same update to constraints, you can factor the constraint updating out into another function like `updateKeyboardConstraints` to avoid duplicating code